### PR TITLE
ci: open the main to dev back-merge pull request automatically

### LINKED ISSUE #561: ci: open the main -> dev back-merge pull request automatically
## Summary

Anything merged directly into `main` — a hotfix, a docs fix, a release bump — leaves `dev` behind until a human remembers to back-merge. Nothing on the GitHub side notices. The drift is silent from the moment it happens until the next release cut trips over it, which can be days later and is diagnosed by whoever happens to be cutting rather than by whoever caused it.

Add a workflow that watches for the divergence and opens the back-merge pull request itself, so the reconciliation is waiting for review instead of waiting to be remembered.

## Why the existing issues do not cover this

Two open issues already address the release path, and this one deliberately does not overlap with either:

- **#528** makes `release-pr` fail when `origin/main` has commits absent from `origin/dev`. That is a gate at the *next release*, so the drift still persists for however long passes before someone cuts one.
- **#532** makes `publish` close the loop for the drift *the release itself creates* (the version bump and the stamped `CHANGELOG` section).

Neither sees a commit that reaches `main` outside the release flow. That is the case here, and it is the more common one: #455 (`rc-main/fix/marketplace-version-lockstep`) and #457 (`docs/beginner-onboarding`) both merged straight into `main` without touching the version, so #528's guard would not have fired and #532's post-`publish` hook would never have run. #530 was the manual cleanup after a release cut, and it is a good picture of the pull request this workflow should be producing on its own.

## Constraints to design against

- **A pull request is mandatory on both branches.** The repository ruleset (`Push Ruleset`, active on `~DEFAULT_BRANCH` and `refs/heads/dev`) requires a pull request with one approving review, and blocks deletion and non-fast-forward pushes. The workflow cannot push to `dev` directly, and it cannot merge its own pull request. That is fine and intended — a human approval is the point. The workflow's job is to make sure the pull request *exists*.
- **A pull request opened with `GITHUB_TOKEN` does not trigger `on: pull_request` workflows.** Left as-is, the back-merge pull request would show zero checks, and CI (`.github/workflows/ci.yml`) is exactly what should be validating the merge result. Resolve this with a GitHub App token or a PAT, or state in the pull request body that checks must be re-run manually. Whichever is chosen, say so explicitly in the workflow comments — a green-looking pull request with no checks is worse than an obviously unchecked one.
- **`Allow GitHub Actions to create and approve pull requests`** must be enabled at the repository or organization level. Confirm before building; if it is off, the workflow needs a token that is not `GITHUB_TOKEN` anyway.
- **`CHANGELOG.md` conflicts are the normal case for release drift**, not the exception: `main` carries the stamped `## [X.Y.Z]` heading while `dev` still has those entries under `## [Unreleased]`. The workflow must handle a conflicting merge as a first-class outcome, not an error path bolted on afterwards.

## Proposed change

One new workflow, `.github/workflows/backmerge.yml`, with two behaviours driven by the same check.

### Triggers

```yaml
on:
  push:
    branches: [main]
  schedule:
    - cron: "0 9 * * *"   # catches drift the push event missed
  workflow_dispatch:
```

The `push` trigger reacts immediately. The schedule is the backstop for anything the push event missed (a workflow failure, a merge that happened while Actions was degraded) and for re-checking a divergence that is still unresolved. `workflow_dispatch` makes it testable without pushing to `main`.

### The check

```bash
git fetch --quiet origin main dev
behind=$(git rev-list --count origin/dev..origin/main)
```

`behind == 0` → nothing to do. Close the tracking issue if one is open, then exit successfully.

### The back-merge attempt

When `behind > 0`, branch off `origin/dev` and merge `origin/main` in. The direction matters: the head branch must contain `dev` plus `main` so the pull request's base can be `dev`. Follow the existing branch convention (`rc-dev/...`, as in #530) and make the name deterministic per `main` head so re-runs are idempotent:

```
rc-dev/chore/back-merge-main-<short-sha-of-origin/main>
```

- **Merge succeeds** → push the branch and open a pull request into `dev`. Title along the lines of `chore: back-merge main into dev (N commit(s))`. The body should list the commits being carried over and name why they are on `main`.
- **Merge conflicts** → abort the merge, push nothing, and fall through to the tracking issue.

If the branch or its pull request already exists for that `main` head, the workflow should leave them alone rather than pushing again — the ruleset forbids non-fast-forward pushes, and rewriting a branch under an open pull request destroys the incremental diff a reviewer is working through.

### The tracking issue

A single issue, found by a stable label (`back-merge` is suggested; it needs creating) rather than by title so the title can carry the live count:

- opened when `behind > 0` and no such issue is open
- updated with a comment when the count changes or a new conflict appears
- closed automatically when `behind` returns to `0`

For the conflict case it should carry copy-pasteable commands and the list of conflicting paths, so resolving it is mechanical:

```bash
git fetch origin main dev
git checkout -b rc-dev/chore/back-merge-main-<sha> origin/dev
git merge origin/main
# resolve, then:
git push -u origin rc-dev/chore/back-merge-main-<sha>
gh pr create --base dev --head rc-dev/chore/back-merge-main-<sha> --title "chore: back-merge main into dev"
```

For a `CHANGELOG.md` conflict specifically, the resolution is stable enough to state once in the issue body: keep `main`'s stamped `## [X.Y.Z]` heading and reopen an empty `## [Unreleased]` above it.

## Explicitly out of scope

- **A `backmerge` subcommand in `scripts/release.py`.** That is #532. This workflow is the net for drift that arrives outside the release flow; #532 removes the recurring source inside it. Both are worth having, and landing only this one means every release still generates a conflicting back-merge pull request that a human resolves by hand.
- **Pushing to `dev` directly, or auto-merging the pull request.** The ruleset forbids the first and requires an approval for the second. Neither should be worked around.
- **Resolving conflicts automatically** (`-X ours`, `-X theirs`, or similar). Both directions silently discard real work. A conflict is a human decision.
- **A local git hook.** The merge happens on GitHub's servers; a client-side hook never sees it and only exists on machines that installed it.

## Done when

- a commit reaching `origin/main` that is absent from `origin/dev` produces, without human action, either a back-merge pull request into `dev` or a tracking issue explaining why one could not be opened
- a clean back-merge opens exactly one pull request, based on `dev`, containing the `main`-only commits
- a conflicting back-merge pushes nothing and opens or updates the tracking issue with the conflicting paths and the commands to resolve them
- re-running the workflow against an unchanged `main` head does not open a second pull request, push over an existing branch, or file a duplicate issue
- the tracking issue closes on its own once `origin/main` and `origin/dev` are back in sync
- the back-merge pull request either carries CI results or says plainly in its body that it does not
- `RELEASE.md` describes the automation and what an operator is expected to do when the tracking issue appears

## Notes for whoever picks this up

- The scheduled run and the push run are the same job with the same logic — resist splitting them.
- Test the conflict path deliberately. The easiest reproduction is a `CHANGELOG.md` edit on both branches touching the `## [Unreleased]` section; that is the shape the real conflicts take.
- If any non-trivial logic lands in Python rather than shell, `.github/workflows/scripts/` is where workflow-owned scripts live (see `lint_skill_refs.py`), and `tests/test_release.py` shows the established pattern for testing git-touching code without shelling out (`_forbid_subprocess`, plus `monkeypatch` on `run`/`capture`).
- The version-shaped guard this complements is at `scripts/release.py:432`, with `_target_ahead_of_main` at `scripts/release.py:326`. The line numbers quoted in #528's body predate later edits; trust the symbol names over the numbers.

### PR DESCRIPTION
**Pull request by mocha06**
## Motivation

A commit that reaches `main` outside the release flow leaves `dev` behind, and nothing on the GitHub side notices.

`release-pr` already refuses to cut while `origin/main` carries commits absent from `origin/dev`. That guard turns a silent hazard into a hard failure, but it fires only at the next release. The drift persists until someone tries to cut one, and it is diagnosed by whoever happens to be cutting rather than by whoever caused it.

The case is live now. `origin/main` carries 9 commits that `origin/dev` does not: the Apache-2.0 LICENSE, the DCO workflow, the conduct and security docs, and the README banner. None of them touched the version, so the version-string guard could not have seen them. Two manual back-merge attempts, #564 and #565, were closed without merging, so this has stood since 1 August and blocks the next release today.

## Outcome

`.github/workflows/backmerge.yml` counts `origin/dev..origin/main` on every push to `main`, daily, and on demand. When the count is zero it closes any open tracking issue and stops. When the count is non-zero it cuts `rc-dev/chore/back-merge-main-<short-sha>` from `origin/dev`, merges `origin/main` in, and opens a pull request into `dev`. The reconciliation waits for review instead of waiting to be remembered.

The decision logic lives in `.github/workflows/scripts/backmerge.py`. Read that file first. The workflow supplies a checkout and credentials, nothing else.

### Behaviour boundaries

- **A conflict is a normal outcome.** The merge is aborted, nothing is pushed, and an issue labelled `back-merge` names the conflicting paths and the commands to resolve them. The run exits zero, because a red run on every release trains people to ignore the signal.
- **Re-runs are idempotent.** The branch name derives from the `main` head, so the same drift always resolves to the same branch. An existing branch or pull request is recognised and left alone. A closed back-merge pull request counts too: closing one is a human decision.
- **Nothing is resolved automatically.** `-X ours` and `-X theirs` both silently discard real work.
- **The tracking issue is one thread.** It is found by label rather than by title, its title carries the live count, it closes on its own once the branches are back in sync, and a re-run that would repeat itself stays silent.
- **The workflow cannot push to `dev` or merge its own pull request.** The ruleset forbids both, and that human approval is the point.

### CI on the back-merge pull request

GitHub does not trigger `on: pull_request` for a pull request opened with `GITHUB_TOKEN`, and `ci.yml` does not match an `rc-dev/*` branch push either. The pull request body states this plainly and says how to run the checks. An unchecked pull request that merely looks green is worse than an obviously unchecked one.

The workflow reads `${{ secrets.BACKMERGE_TOKEN || secrets.GITHUB_TOKEN }}`. Setting that secret turns checks on with no change to the workflow. Use a GitHub App token, or a fine-grained PAT scoped to this repository with `contents: write`, `pull-requests: write`, and `issues: write`. A classic PAT is the wrong instrument for a public repository: it is user-scoped, carries access to every repository its owner can reach, and does not expire.

Secrets stay out of reach of forks. GitHub withholds them from `pull_request` runs raised by a fork, and this workflow has no `pull_request_target` or `workflow_run` trigger, which are the two patterns that leak them.

### Two assumptions in #561 that did not hold

1. The issue calls `CHANGELOG.md` conflicts "the normal case". For the live drift, `CHANGELOG.md` auto-merges. The single conflict is binary: `docs/images/pipefy-developers-banner.png`, changed independently on both branches. A binary conflict has no textual resolution to suggest, so the tracking issue names the conflicting paths generically and adds the CHANGELOG recipe only when `CHANGELOG.md` is actually among them.
2. The issue asks for a tracking issue whenever the count is non-zero, while its Done-when section asks for "either a back-merge pull request into `dev` **or** a tracking issue explaining why one could not be opened". Opening both for one drift is noise, so the issue is filed only when a pull request could not be opened. Done-when is the acceptance criteria, so it wins.

The issue also describes #528 as pending. It landed in #572, which strengthens the case here rather than changing it: drift now blocks a release outright instead of quietly mis-stamping the notes.

## Verification

Scenarios 1 to 5 ran against the live repository, in the dry-run mode the `pull_request` trigger uses. The rest are unit tests, because forcing them live would mean pushing branches and filing issues to observe them.

| # | Scenario | Expected | Observed | Verdict |
|---|---|---|---|---|
| 1 | Real repo state, `main` 9 commits ahead | Count the drift, name the head | `origin/main is 9 commit(s) ahead at 4691d13` | Pass |
| 2 | Binary conflict on the banner PNG | Abort, push nothing, route to the tracking issue | Conflict on `docs/images/pipefy-developers-banner.png`; no push; issue body produced | Pass |
| 3 | Conflict guidance for a non-CHANGELOG conflict | No CHANGELOG recipe | Issue body names the PNG alone; no `## [Unreleased]` advice | Pass |
| 4 | Exit code on the conflict path | `0` | `0` | Pass |
| 5 | Dry run performs no remote write | Every `gh` and `push` announced, not run | All announced under `[dry run]` | Pass |
| 6 | `rev-list` range direction | `origin/dev..origin/main`; reversed, it would propose a back-merge on every commit landing on `dev` | Unit test pins the exact argv | Pass |
| 7 | Fetch refspec | Explicit `+refs/heads/…:refs/remotes/origin/…`; a bare fetch under `actions/checkout`'s narrow refspec never creates `refs/remotes/origin/dev` | Unit test pins the exact argv | Pass |
| 8 | Conflicting paths read before `merge --abort` | Paths collected first; abort clears the unmerged index | Unit test asserts the ordering | Pass |
| 9 | A pull request already covers this `main` head | Leave it alone, in any state | Unit test | Pass |
| 10 | Branch pushed, no pull request | Open the pull request, do not re-push | Unit test | Pass |
| 11 | `gh pr create` rejected | Fall through to the tracking issue, naming the cause | Unit test | Pass |
| 12 | Branches back in sync | Close the tracking issue | Unit test | Pass |
| 13 | Re-run with an unchanged state | No duplicate comment | Unit test on the state marker | Pass |
| 14 | Conflicting paths in a different order | Same marker, still silent | Unit test | Pass |
| 15 | Merge commit sign-off | DCO trailer present, matching the configured identity | Unit test | Pass |

Gates: `pytest -m "not integration"` (4196 passed, 5 skipped), `ruff check`, `ruff format --check`, `lint-imports`, `bump_version.py verify`, `lint_skill_refs.py`, and the skills-frontmatter check. The last two run because this diff touches `.github/workflows/scripts/**`, which the skills-lint path filter matches.

## Docs

- `RELEASE.md`: a new "Keeping `dev` and `main` reconciled" section, a cross-reference from the `release-pr` precondition, and a row in the automation reference table.
- `CHANGELOG.md`: an entry under `## [Unreleased]`.

Closes #561.

### CODE REVIEW

File: .github/workflows/scripts/backmerge.py [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+    reason: str,
+) -> None:
+    """Open the tracking issue, or comment on the open one when the state changed."""
+    ensure_label()
+    marker = state_marker(sha, behind, conflicts)
+    body = issue_body(sha, behind, branch, conflicts, commits, marker, run_url, reason)
+    title = issue_title(behind)
+    issue = open_tracking_issue()
+    if not issue:
+        mutate(
```
Comment (adriannoes) [inline .github/workflows/scripts/backmerge.py]: **Blocking.** `track()` calls `mutate(...)` for `gh issue create` / comment / edit and ignores the exit code. On the conflict path (and on `open_pr` failure) `main()` then always `return 0`.

Exit 0 on conflict is fine when the `back-merge` issue actually lands. That issue is the compensating signal this workflow chose instead of a red X. If label/issue create also fails, Actions stays green with no PR and no issue, which is the silent drift this automation exists to surface.

I would make compensating writes fail-closed: if conflict or `gh pr create` failed and issue create/comment did not succeed, exit non-zero. Keep exit 0 only when the PR or the tracking issue actually landed. Add a unit test that stubs `mutate` to return 1 on issue create and asserts a non-zero exit.

Done when:

- conflict / PR-reject path exits non-zero when tracking-issue writes fail
- a regression test fails if that check is removed

File: .github/workflows/scripts/backmerge.py [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+            track(sha, behind, branch, [], commits, run_url, "`gh pr create` failed.")
+        return 0
+
+    conflicts = attempt_merge(branch, merge_message(behind, signoff_trailer()))
+    if conflicts:
+        print(f"Merge conflicts in {len(conflicts)} path(s); pushing nothing.")
+        track(sha, behind, branch, conflicts, commits, run_url, "")
+        return 0
+
+    mutate(["git", "push", "origin", branch])
```
Comment (adriannoes) [inline .github/workflows/scripts/backmerge.py]: **Optional.** `mutate(["git", "push", "origin", branch])` discards its exit code. A failed push plus a failed `gh pr create` still opens a tracking issue that says the branch is pushed and ready.

I would check the push exit before `open_pr`, and on failure `track` with a push-failed reason (skip the Actions-PR-permission copy). A clean-merge success test that asserts the push argv would also close the H5 gap. Your call whether to take this before merge.

File: RELEASE.md [hunk 3]
```diff
@@ -129,6 +129,21 @@ Two limits worth knowing rather than being surprised by:
...
 - **The gate proves process startup, not behaviour.** A dependency release that breaks behaviour without breaking imports passes it, and the suite would not see it either, since the suite runs against the lock. Closing that band means running some portion of the tests unlocked.
 
+## Keeping `dev` and `main` reconciled
+
+`release-pr` refuses to cut while `origin/main` carries commits absent from `origin/dev`. That guard turns a silent hazard into a hard failure, but only at the *next* release — the drift itself persists for however long passes before someone tries to cut one, and it is diagnosed by whoever happens to be cutting rather than by whoever caused it.
+
+`.github/workflows/backmerge.yml` closes that window from the other side. On every push to `main`, daily, and on demand, it counts `origin/dev..origin/main`. When that is zero it does nothing. When it is not, it cuts `rc-dev/chore/back-merge-main-<short-sha>` from `origin/dev`, merges `origin/main` into it, and opens a pull request into `dev` — so the reconciliation is waiting for review instead of waiting to be remembered.
```
Comment (adriannoes) [inline RELEASE.md]: **Optional.** This sentence says a zero count "does nothing," but `main()` closes any open `back-merge` tracking issue when the branches are back in sync (and the same section later says the issue closes on its own). I would say it closes any open tracking issue and otherwise stops. Your call.

**Review by adriannoes — CHANGES_REQUESTED** [review]
Reviewed at `4080e8f`: local backmerge unit suite (24), ruff, CI green including the path-filtered dry-run of this workflow; security static pass; no MCP/CLI surface.

**Verdict: needs changes** into `dev`.

### Required before merge

- Fail-closed when the compensating `back-merge` issue cannot be written after conflict or `gh pr create` failure. See the thread on `.github/workflows/scripts/backmerge.py:382`.

### Decisions I would make

- Prefer merging #577 (live banner conflict back-merge) before treating release cuts as unblocked; #576 arms future detection and stays inert for `push`/`schedule` until the workflow reaches default `main`.
- `Closes #561` will not auto-close on a merge to `dev` (default branch is `main`). Close #561 manually once the workflow is on `main`, or carry the keyword through promotion.
- If a conflict issue is filed and a later tip opens a clean back-merge PR, the old issue stays open until `behind==0`. I would close or comment when `open_pr` succeeds so conflict guidance does not outlive the PR.

### Optional

Your call on all of these.

- Check `git push` exit and stop claiming the branch is ready when it is not (pointer on `backmerge.py:498`).
- Align `RELEASE.md` zero-count prose with the close-issue behavior (pointer on `RELEASE.md:136`).

### What worked well

- Decision logic in unit-tested Python with a thin workflow shell; conflict-as-normal, idempotent branch naming, closed-PR respect, and no auto-resolve are coherent.
- Dry-run on `pull_request` already exercised the live binary conflict path without mutating the remote.

<details>
<summary>Review path</summary>

Tier `standard`, trust `trusted`, reviewed at `05ed348e...4080e8f2`.

| Gate | Result |
| :--- | :--- |
| phase3 pytest tests/test_backmerge.py | 24 passed |
| CI | green (backmerge job = dry-run) |
| H15 security | clean |
| H16 live MCP/CLI | n/a |

</details>

**Comment by mocha06** [comment]
Pushed `5a9e269` — all three findings taken, plus two more that tracing them turned up.

## Applied

**Blocking — fail-closed on the compensating write.** Agreed, and it is the right framing: exiting zero on a conflict is defensible only while the tracking issue *is* the signal. `track()` now returns whether the drift got recorded, and `main()` exits non-zero when neither a pull request nor an issue landed. A failed `gh issue edit --title` stays cosmetic and does not turn the run red, since the drift is recorded either way. Covered by `test_a_failed_tracking_issue_write_fails_the_run`, `test_a_failed_issue_comment_fails_the_run`, and `test_a_title_edit_failure_does_not_fail_the_run`.

**Push exit code.** Taken. The push exit is checked before `open_pr`, with its own reason saying nothing reached the remote. The `reason` argument now stands alone in the issue body instead of being wrapped in one hardcoded "what failed was opening the pull request", so each caller says what actually happened. `test_a_failed_push_does_not_claim_the_branch_is_ready` asserts the old "pushed and ready" copy cannot come back, and `test_a_clean_merge_pushes_then_opens_the_pull_request` pins the push argv and ordering — the H5 gap you pointed at.

**Stale issue outliving a clean PR.** Taken. A successful `open_pr` closes the open tracking issue with a comment naming the branch.

**`RELEASE.md` zero-count prose.** Fixed — it now says it closes any open tracking issue and stops.

## Two more, found while tracing the third

**A closed pull request silently swallowed the drift.** `existing_pr` searched `--state all` and `main()` returned 0 on any match, so a pull request closed without merging meant no pull request *and* no issue, forever. That is the observed failure, not a hypothetical: #564 and #565 were both closed unmerged and the divergence sat unnoticed for days. It violated the issue's own Done-when. A closed one is still neither reopened nor re-pushed — that decision stands — but the drift is now recorded.

**Squash-merging a back-merge makes the whole thing non-convergent.** #577 was squash-merged. The second parent was discarded, so `dev` has every byte of `main`'s content and none of its ancestry: `git rev-list --count origin/dev..origin/main` is **still 9**, `release-pr` is still blocked, and this workflow would have proposed the same back-merge on every run forever. The generated pull request body now requires "Create a merge commit" and explains why, and a merged pull request found alongside surviving drift is diagnosed as the squash specifically, with the commands to redo it. `RELEASE.md` states it too.

The live dry run against current `main` produces exactly that diagnosis, naming #577.

## Not changed

`Closes #561` stays. Carrying it through promotion is how it fires, same as #528 via #572 — no auto-close on a `dev` merge either way.

Gates: 4207 passed / 5 skipped, `ruff check`, `ruff format --check`, `lint-imports`, `bump_version.py verify`, `lint_skill_refs.py`. Back-merge suite is 35 tests, up from 24.

**Review by adriannoes — APPROVED** [review]
Re-checked at `5a9e269e` against the prior REQUEST_CHANGES on `4080e8f2`.

The fix commit makes `track()` report whether the drift was recorded and fails the run when neither a pull request nor the tracking issue landed. It also checks the `git push` exit, stops claiming a missing branch is ready, closes a superseded conflict issue when a clean PR opens, and aligns the RELEASE zero-count prose with the close-issue behavior. Local `tests/test_backmerge.py` is 35/35 green; CI on this tip is green.

**Verdict: merge-ready** into `dev`.

Prior blocking and optional threads are answered by this tip. No further code changes required from that review.
